### PR TITLE
Add unit test content for extracting functions 

### PR DESCRIPTION
**Pull request by ProchaLu**
PR for:
- https://github.com/upleveled/courses/pull/2062

Add Jest unit test content for extracting functions from existing code. This also adds `birthDate` to the single animals page and shows the student how to handle dates. This update requires us to update multiple lectures as follows:

### Update Next.js lecture

1. Add a `birthDate` property to the animals array of objects
```js
  {
    id: 1,
    firstName: 'Lucia',
    type: 'Lion',
    accessory: 'Car',
    birthDate: new Date('2020-12-25'),
  },
```

2. Add the `birthDate` property and the calculation of the days until the next birthday to the `animals` single page
```js
  const currentDate = new Date();

  // Create new date object to avoid mutating the original birth date
  const nextBirthDate = new Date(singleAnimal.birthDate);

  // Set birth date year to current year
  nextBirthDate.setUTCFullYear(currentDate.getFullYear());

  // Set UTC time to 0 to compare only days (avoid time zones)
  currentDate.setUTCHours(0, 0, 0, 0);
  nextBirthDate.setUTCHours(0, 0, 0, 0);

  if (nextBirthDate.getTime() < currentDate.getTime()) {
    nextBirthDate.setUTCFullYear(currentDate.getFullYear() + 1);
  }

  const daysUntilNextBirthDay =
    (nextBirthDate.getTime() - currentDate.getTime()) / (1000 * 60 * 60 * 24);
```

3. Display the birth date and the remaining days until the next birthday on the `animals` single page
```js
      <div>
        Birth date:{' '}
        {singleAnimal.birthDate.toLocaleDateString('en-GB', {
          year: 'numeric',
          month: '2-digit',
          day: '2-digit',
        })}
      </div>
      <div>Days left until birthday: {daysUntilBirthday}</div>
```
Screenshot 2023-12-07 at 13 49 53

### Update Database and Migrations lecture

1. Update the migrations file to create a table for animals with a `birthDate` property
```js
export async function up(sql: Sql) {
  await sql`
    CREATE TABLE animals (
      id integer PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
      first_name varchar(30) NOT NULL,
      type varchar(30) NOT NULL,
      accessory varchar(30),
      birth_date date NOT NULL
    );
  `;
}
``` 

2. Update the migrations file to insert the animals with a `birthDate` property 
```js
export async function up(sql: Sql) {
  for (const animal of animals) {
    await sql`
      INSERT INTO animals
        (first_name, type, accessory, birth_date)
      VALUES
        (${animal.firstName}, ${animal.type}, ${animal.accessory}, ${animal.birthDate})
  `;
  }
}
```
Screenshot 2023-10-16 at 10 34 57

### Update testing lecture 

1. Extract and create 2 new functions for the Jest test and move it to a new `util/date.ts` file
```js
export function formatDate(
  date: Date,
  locale: string = 'en-GB',
  options: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  },
) {
  if (!(date instanceof Date) || isNaN(date.getTime())) {
    throw new Error('Pass only dates!');
  }
  return date.toLocaleDateString(locale, options);
}

export function getDaysUntilNextBirthday(currentDate: Date, birthDate: Date) {
  if (!(currentDate instanceof Date && birthDate instanceof Date)) {
    throw new Error('Pass only dates!');
  }

  if (birthDate.getTime() > currentDate.getTime()) {
    throw new Error('Birth date must be before current date!');
  }

  // Set UTC time to 0 to compare only days (avoid time zones)
  currentDate.setUTCHours(0, 0, 0, 0);
  birthDate.setUTCHours(0, 0, 0, 0);

  birthDate.setUTCFullYear(currentDate.getFullYear());

  if (birthDate.getTime() < currentDate.getTime()) {
    birthDate.setUTCFullYear(currentDate.getFullYear() + 1);
  }

  const daysUntilNextBirthday =
    (birthDate.getTime() - currentDate.getTime()) / (1000 * 60 * 60 * 24);

  return daysUntilNextBirthday;
}
```

2. Create 2 new Jest tests for the functions `formatDate()` and `calculateDaysUntilNextBirthday()`

### CODE REVIEW

**Comment by ProchaLu** [comment]
Update for the Learning Platform calendar
- https://github.com/upleveled/courses/pull/2062

File: app/animals/[animalId]/page.tsx [hunk 1]
```diff
@@ -23,10 +23,30 @@ export default async function AnimalPage(props: Props) {
...
 
+  const formattedBirthDate = singleAnimal.birthDate.toLocaleDateString('de-AT');
+
+  const currentDate = new Date();
+
+  const nextBirthDate = new Date(
+    currentDate.getFullYear(),
+    singleAnimal.birthDate.getMonth(),
+    singleAnimal.birthDate.getDate(),
+  );
```
Comment (karlhorky) [inline app/animals/[animalId]/page.tsx:34]: does this fail if the birthday has already passed?

File: migrations/00001-insertAnimals.ts [hunk 2]
```diff
@@ -1,20 +1,50 @@
...
 
 const animals = [
-  { id: 1, firstName: 'lucia', type: 'Lion', accessory: 'Car' },
-  { id: 2, firstName: 'macca', type: 'Dog', accessory: 'Comb' },
-  { id: 3, firstName: 'jojo', type: 'Dodo', accessory: 'Dojo' },
-  { id: 4, firstName: 'flo', type: 'Parrot', accessory: 'carrot' },
-  { id: 5, firstName: 'bili', type: 'Capybara', accessory: 'Pen' },
+  {
+    id: 1,
+    firstName: 'lucia',
```
Comment (karlhorky) [inline migrations/00001-insertAnimals.ts]: let's capitalize the names

File: util/__tests__/birthDate.test.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { expect, test } from '@jest/globals';
+import { calculateDaysUntilBirthday } from '../date';
+
+const currentDate = new Date('2023-10-14');
+const birthDate = new Date('2021-11-25');
+
+test('calculate days until birthday', () => {
+  expect(calculateDaysUntilBirthday(currentDate, birthDate)).toBe(42);
+});
```
Comment (karlhorky) [inline util/__tests__/birthDate.test.ts]: let's try to hit at least 30 lines in this file or so. add:

- more test cases (more data)
- tests for edge cases

File: util/__tests__/dateformatter.test.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { expect, test } from '@jest/globals';
+import { formatDate } from '../date';
+
+const currentDate = new Date('2023-10-14');
+
+test('format date for displaying the birth date', () => {
+  expect(formatDate(currentDate)).toBe('14.10.2023');
+});
```
Comment (karlhorky) [inline util/__tests__/dateformatter.test.ts]: let's try to hit at least 30 lines in this file or so. add:

- more test cases (more data)
- tests for edge cases

File: migrations/00001-insertAnimals.ts (hunk 2 above)
Comment (ProchaLu) [reply migrations/00001-insertAnimals.ts]: Capitalize the names
[495d035](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/495d035436e04adbc32cf9e92d54bb9f491196b9)

File: app/animals/[animalId]/page.tsx [hunk 5]
```diff
@@ -23,10 +23,30 @@ export default async function AnimalPage(props: Props) {
...
+
+  if (nextBirthDate < currentDate) {
+    nextBirthDate.setFullYear(currentDate.getFullYear() + 1);
+  }
+
+  const daysUntilBirthday = Math.ceil(
+    (nextBirthDate.getTime() - currentDate.getTime()) / (1000 * 60 * 60 * 24),
+  );
```
Comment (karlhorky) [inline app/animals/[animalId]/page.tsx]: why is all of this code here? shouldn't it use the function here instead?

File: app/animals/[animalId]/page.tsx [hunk 6]
```diff
@@ -23,10 +23,30 @@ export default async function AnimalPage(props: Props) {
     return notFound();
   }
 
+  const formattedBirthDate = singleAnimal.birthDate.toLocaleDateString('de-AT');
```
Comment (karlhorky) [inline app/animals/[animalId]/page.tsx]: this could be inlined below to show that you can extract a function from an inline JSX expression too

File: util/date.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+export function formatDate(date: Date) {
+  return date.toLocaleDateString('de-AT');
```
Comment (karlhorky) [inline util/date.ts]: can you use more options of `toLocaleDateString()` here to make the function more valuable?

File: util/date.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+export function calculateDaysUntilBirthday(currentDate: Date, birthDate: Date) {
+  const today = new Date();
+  const birthday = new Date(birthDate);
+
+  const nextBirthday = new Date(
+    today.getFullYear(),
+    birthday.getMonth(),
+    birthday.getDate(),
+  );
```
Comment (karlhorky) [inline util/date.ts]: does this fail if the birthday for the current year has already passed?

**Comment by karlhorky** [comment]
- [x] title could use some work to improve clarity
  - [x] "Update Next.js and Jest lectures to handle dates" sounds like "dates were there already there before, but unhandled"
  - [x] we don't need to mention "lecture", since this project doesn't need to know about the lectures

File: migrations/00001-insertAnimals.ts (hunk 2 above)
Comment (ProchaLu) [reply migrations/00001-insertAnimals.ts]: Capitalized animal names
[495d035](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/495d035436e04adbc32cf9e92d54bb9f491196b9)

File: app/animals/[animalId]/page.tsx (hunk 5 above)
Comment (ProchaLu) [reply app/animals/[animalId]/page.tsx]: I thought Victor wanted to check it and how he should implement this function.
Deleted the code and used the extracted function
[179c9da](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/179c9da845f6f7c4e83787580913dbd68494e7e8)

File: app/animals/[animalId]/page.tsx (hunk 6 above)
Comment (ProchaLu) [reply app/animals/[animalId]/page.tsx]: I thought Victor wanted to check it and how he should implement this function.
Deleted the code and used the extracted function
[179c9da](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/179c9da845f6f7c4e83787580913dbd68494e7e8)

File: util/date.ts (hunk 8 above)
Comment (ProchaLu) [reply util/date.ts]: In this `if`, if the birthday already passed, I change the year to the following year and calculate the days until this new date:
```js
  if (nextBirthday < today) {
    nextBirthday.setFullYear(today.getFullYear() + 1);
  }
```

File: app/animals/[animalId]/page.tsx (hunk 5 above)
Comment (ProchaLu) [reply app/animals/[animalId]/page.tsx]: Refactored and simplified the function in this commit [7725e81](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/7725e818630f1f788506f79bc7b0f758b456cf29)

```js
export function calculateDaysUntilNextBirthday(
  currentDate: Date,
  birthDate: Date,
) {
  const currentYear = currentDate.getFullYear();
  const birthDateThisYear = new Date(birthDate);
  birthDateThisYear.setFullYear(currentYear);

  const daysDifference = birthDateThisYear.getTime() - currentDate.getTime();
  const days = Math.ceil(daysDifference / (1000 * 60 * 60 * 24));

  return days < 0 ? days + 365 : days;
}
```

File: util/date.ts (hunk 7 above)
Comment (ProchaLu) [reply util/date.ts]: Made the `formatDate()` function more valuable with additional options and locale settings
Commit [7725e81](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/7725e818630f1f788506f79bc7b0f758b456cf29)
```js
export function formatDate(
  date: Date,
  locale: string,
  options: Intl.DateTimeFormatOptions = {},
) {
  return date.toLocaleDateString(locale, options);
}
```

File: util/date.ts [hunk 9]
```diff
@@ -10,12 +10,17 @@ export function calculateDaysUntilNextBirthday(
...
-  const daysDifference = birthDateThisYear.getTime() - currentDate.getTime();
-  const days = Math.ceil(daysDifference / (1000 * 60 * 60 * 24));
+  const daysDifference = Math.ceil(
+    (birthDateThisYear.getTime() - currentDate.getTime()) /
+      (1000 * 60 * 60 * 24),
+  );
 
-  return days < 0 ? days + 365 : days;
+  if (daysDifference === -0) {
+    return 0;
```
Comment (ProchaLu) [inline util/date.ts]: To handle the case where the birthday is on the same day as the current date, I added an early return to return `0` instead of `-0`. This is because `Math.ceil` returns `-0` when the input is `-0`

File: util/__tests__/dateformatter.test.ts (hunk 4 above)
Comment (ProchaLu) [reply util/__tests__/dateformatter.test.ts]: Added more test cases
[72ce239](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/72ce2392ad3353ee345aaf72d6fb0e654ac25af6)

File: util/__tests__/birthDate.test.ts (hunk 3 above)
Comment (ProchaLu) [reply util/__tests__/birthDate.test.ts]: Added more test cases
[72ce239](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/72ce2392ad3353ee345aaf72d6fb0e654ac25af6)

File: util/date.ts (hunk 9 above)
Comment (ProchaLu) [reply util/date.ts]: Deleted the `if` block, and now use `Math.abs()` 
[8d8a5dd](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/8d8a5dd6c0eb81213f19aa99572d4d7d2ae426c9)

File: util/date.ts (hunk 8 above)
Comment (ProchaLu) [reply util/date.ts]: Removed this `if` block, and changed the birth date function.
```js
export function calculateDaysUntilNextBirthday(
  currentDate: Date,
  birthDate: Date,
) {
  if (!(currentDate instanceof Date && birthDate instanceof Date)) {
    throw new Error('Pass only dates!');
  }
  const birthDateThisYear = new Date(birthDate);
  birthDateThisYear.setFullYear(currentDate.getFullYear());

  const daysDifference = Math.ceil(
    (birthDateThisYear.getTime() - currentDate.getTime()) /
      (1000 * 60 * 60 * 24),
  );

  return Math.abs(daysDifference < 0 ? daysDifference + 365 : daysDifference);
}```

File: app/animals/[animalId]/page.tsx (hunk 5 above)
Comment (ProchaLu) [reply app/animals/[animalId]/page.tsx]: refactored again and now use `Math.abs()` to  avoid the issue with `-0` 
[8d8a5dd](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/8d8a5dd6c0eb81213f19aa99572d4d7d2ae426c9)

File: util/date.ts [hunk 10]
```diff
@@ -10,6 +10,9 @@ export function calculateDaysUntilNextBirthday(
   currentDate: Date,
   birthDate: Date,
 ) {
+  if (!(currentDate instanceof Date && birthDate instanceof Date)) {
+    throw new Error('Pass only dates!');
+  }
```
Comment (karlhorky) [inline util/date.ts]: nice!

should it also have a 2nd error block when the `currentDate` is before the `birthDate`? (and test cases for this)

File: util/date.ts [hunk 11]
```diff
@@ -18,9 +21,5 @@ export function calculateDaysUntilNextBirthday(
       (1000 * 60 * 60 * 24),
   );
 
-  if (daysDifference === -0) {
-    return 0;
-  }
-
-  return daysDifference < 0 ? daysDifference + 365 : daysDifference;
+  return Math.abs(daysDifference < 0 ? daysDifference + 365 : daysDifference);
```
Comment (karlhorky) [inline util/date.ts]: - where does the 365 come from?
  - maybe this can somehow be worked into the calculation above?
  - if you're doing that, maybe also the `Math.abs` can be moved into the calculation above?
  - if you do [this suggestion](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/files#r1368878289) (throw error if the `currentDate` is before the `birthDate`), maybe you don't need some / both of these things

File: util/__tests__/dateformatter.test.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    }),
+  ).toBe('20.10.23');
+});
+
+test('throws an error when dates are not valid', () => {
+  expect(formatDate(new Date('25-03-2023'), 'de-DE')).toBe('Invalid Date');
+
+  expect(formatDate(new Date('25.03.2023'), 'de-DE')).toBe('Invalid Date');
+
+  expect(formatDate(new Date('1698054125000'), 'de-DE')).toBe('Invalid Date');
```
Comment (karlhorky) [inline util/__tests__/dateformatter.test.ts]: shouldn't this be something like `expect(drinkOctopus).toThrow('yuck');` ?

https://jestjs.io/docs/expect#tothrowerror:~:text=expect(drinkOctopus).toThrow(%27yuck%27)%3B

I guess our function should probably throw something here?

File: util/date.ts (hunk 7 above)
Comment (karlhorky) [reply util/date.ts]: nice, but you're just passing the options through - what I meant was to hardcode at least some of the option values inside the function so that our function is more valuable than just calling `date.toLocaleDateString(...)`

(right now, it's just a wrapper for it which doesn't actually add any features (it would be better to get rid of this wrapper, because it doesn't help with programming, except maybe remembering the name `formatDate` being... easier or something))

File: util/__tests__/dateformatter.test.ts
Comment (karlhorky) [inline util/__tests__/dateformatter.test.ts:1]: name of this file should match the other one, shouldn't it?

eg. `util/dates.ts` and `util/__tests__/dates.test.ts`

File: util/date.ts
Comment (karlhorky) [inline util/date.ts:1]: would recommend `util/dates.ts` instead, sounds a bit better

File: util/date.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
+export function formatDate(
+  date: Date,
+  locale: string,
+  options: Intl.DateTimeFormatOptions = {},
+) {
+  return date.toLocaleDateString(locale, options);
+}
+
+export function calculateDaysUntilNextBirthday(
```
Comment (karlhorky) [inline util/date.ts:9]: we can probably drop "calculate" from this function name, many functions are for calculating

File: playwright/navigation.spec.ts [hunk 14]
```diff
@@ -1,11 +1,11 @@
...
-  { id: 2, firstName: 'macca', type: 'Dog', accessory: 'Comb' },
-  { id: 3, firstName: 'jojo', type: 'Dodo', accessory: 'Dojo' },
-  { id: 4, firstName: 'flo', type: 'Parrot', accessory: 'carrot' },
-  { id: 5, firstName: 'bili', type: 'Capybara', accessory: 'Pen' },
+  { id: 1, firstName: 'Lucia', type: 'Lion', accessory: 'Car' },
+  { id: 2, firstName: 'Macca', type: 'Dog', accessory: 'Comb' },
+  { id: 3, firstName: 'Jojo', type: 'Dodo', accessory: 'Dojo' },
+  { id: 4, firstName: 'Flo', type: 'Parrot', accessory: 'carrot' },
+  { id: 5, firstName: 'Bili', type: 'Capybara', accessory: 'Pen' },
 ];
```
Comment (karlhorky) [inline playwright/navigation.spec.ts]: - should this array be moved down to directly above the `for..of`? (in a separate PR)
- `carrot` is not capital case here (maybe the casing should be moved to a different PR?)
- should this include dates too? (eg. is it copied from the fake database at `database/animals.ts`?)
- should `database/animals.ts` also be updated? https://github.com/upleveled/next-js-example-fall-2023-atvie/blob/db730cda672140dc61bb46faf88b4695054a8ddf/database/animals.ts#L10-L16

File: playwright/navigation.spec.ts [hunk 15]
```diff
@@ -1,11 +1,11 @@
...
 const animals = [
-  { id: 1, firstName: 'lucia', type: 'Lion', accessory: 'Car' },
-  { id: 2, firstName: 'macca', type: 'Dog', accessory: 'Comb' },
-  { id: 3, firstName: 'jojo', type: 'Dodo', accessory: 'Dojo' },
-  { id: 4, firstName: 'flo', type: 'Parrot', accessory: 'carrot' },
-  { id: 5, firstName: 'bili', type: 'Capybara', accessory: 'Pen' },
+  { id: 1, firstName: 'Lucia', type: 'Lion', accessory: 'Car' },
+  { id: 2, firstName: 'Macca', type: 'Dog', accessory: 'Comb' },
+  { id: 3, firstName: 'Jojo', type: 'Dodo', accessory: 'Dojo' },
+  { id: 4, firstName: 'Flo', type: 'Parrot', accessory: 'carrot' },
```
Comment (karlhorky) [inline playwright/navigation.spec.ts:7]: `carrot` is not capital case here

File: util/date.ts
Comment (ProchaLu) [reply util/date.ts:1]: changed filename in this commit [d936c1e](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/d936c1efd5e3ba9b1cd3b7768cd56780d4a8d579)

File: util/__tests__/dateformatter.test.ts
Comment (ProchaLu) [reply util/__tests__/dateformatter.test.ts:1]: changed filename in this commit [d936c1e](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/d936c1efd5e3ba9b1cd3b7768cd56780d4a8d579)

File: util/__tests__/dateformatter.test.ts (hunk 12 above)
Comment (ProchaLu) [reply util/__tests__/dateformatter.test.ts]: Throw an error now before the `date.toLocaleDateString` returns with `Invalid Date`
```js
  if (!(date instanceof Date) || isNaN(date.getTime())) {
    throw new Error('Pass only dates!');
  }
```

Commit [d936c1e](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/d936c1efd5e3ba9b1cd3b7768cd56780d4a8d579)

File: util/date.ts (hunk 10 above)
Comment (ProchaLu) [reply util/date.ts]: Added a second error block 
```js
  if (birthDate.getTime() > currentDate.getTime()) {
    throw new Error('Birth date must be before current date!');
  }
```

Commit [d936c1e](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/d936c1efd5e3ba9b1cd3b7768cd56780d4a8d579)

File: util/date.ts (hunk 11 above)
Comment (ProchaLu) [reply util/date.ts]: I was able to remove the `365` which causes bugs in a leap year. Now I changed the function so I don't have to use `Math.abs` and it should work as intended.

```js
export function getDaysUntilNextBirthday(currentDate: Date, birthDate: Date) {
  if (!(currentDate instanceof Date && birthDate instanceof Date)) {
    throw new Error('Pass only dates!');
  }

  if (birthDate.getTime() > currentDate.getTime()) {
    throw new Error('Birth date must be before current date!');
  }

  birthDate.setFullYear(currentDate.getFullYear());

  if (birthDate.getTime() < currentDate.getTime()) {
    birthDate.setFullYear(currentDate.getFullYear() + 1);
  }

  const daysUntilNextBirthday =
    birthDate.getDate() === currentDate.getDate() &&
    birthDate.getMonth() === currentDate.getMonth()
      ? 0
      : Math.ceil(
          (birthDate.getTime() - currentDate.getTime()) / (1000 * 60 * 60 * 24),
        );

  return daysUntilNextBirthday;
}
```

File: util/date.ts (hunk 7 above)
Comment (ProchaLu) [reply util/date.ts]: Changed the function, now it should be more valuable 
```js
export function formatDate(
  date: Date,
  locale: string = 'de-DE',
  options: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  },
) {
  if (!(date instanceof Date) || isNaN(date.getTime())) {
    throw new Error('Pass only dates!');
  }
  return date.toLocaleDateString(locale, options);
}
```

Commit [d936c1e](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/d936c1efd5e3ba9b1cd3b7768cd56780d4a8d579)

File: app/animals/[animalId]/page.tsx [hunk 16]
```diff
@@ -24,7 +24,7 @@ export default async function AnimalPage(props: Props) {
     return notFound();
   }
 
-  const daysUntilNextBirthday = calculateDaysUntilNextBirthday(
+  const daysLeftUntilBirthday = daysUntilNextBirthday(
```
Comment (karlhorky) [inline app/animals/[animalId]/page.tsx]: I actually deleted the original comment where I said "remove the `calculate` from the function name" because of situations like this - naming the `const` variable is difficult because it collides with the function name

the const seems like it should stay as `daysUntilNextBirthday`

so maybe best to keep the `calculate` prefix - or use some synonym like `get` or `compute`

File: util/__tests__/dateformatter.test.ts
Comment (karlhorky) [reply util/__tests__/dateformatter.test.ts:1]: there is still a file called `util/__tests__/dateformatter.test.ts`

Screenshot 2023-10-24 at 11 50 20

File: util/__tests__/dateformatter.test.ts [hunk 17]
```diff
@@ -1,10 +1,12 @@
 import { expect, test } from '@jest/globals';
-import { formatDate } from '../date';
+import { formatDate } from '../dates';
 
 test('format date for displaying the date with different locale date formats', () => {
-  expect(formatDate(new Date('2023-10-20'), 'us-US')).toBe('10/20/2023');
+  expect(formatDate(new Date('2023-10-20'))).toBe('20.10.2023');
+
+  expect(formatDate(new Date('2023-10-20'))).not.toBe('10/20/2023');
```
Comment (karlhorky) [inline util/__tests__/dateformatter.test.ts]: - hm, does this make sense, where we have the previous assertion?
- I do agree that we should try to show different matchers eg. a `.not...` matcher, but maybe we should do it in a different place that is more of a common way to do it?

File: util/__tests__/dateformatter.test.ts
Comment (ProchaLu) [reply util/__tests__/dateformatter.test.ts:1]: Removed the file and added the tests to the `dates.test.ts` file
Commit [1aef4b9](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/1aef4b9a6007d1c0cf96bdaf5ff1f6ab9c0131c7)

File: app/animals/[animalId]/page.tsx (hunk 16 above)
Comment (ProchaLu) [reply app/animals/[animalId]/page.tsx]: I would use the `get` synonym instead of `calculate` because `get` tells me that I get a value, when using `calculate` it sounds more like an action instead. 
Commit [1aef4b9](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/1aef4b9a6007d1c0cf96bdaf5ff1f6ab9c0131c7)

File: util/__tests__/dateformatter.test.ts (hunk 17 above)
Comment (ProchaLu) [reply util/__tests__/dateformatter.test.ts]: Removed the `.not.toBe` test because it is the same as the previous assertion. Used the `.not.toBe` now in a different test where I assert that the function is not throwing an error.
```js
expect(() => formatDate(new Date())).not.toThrow('Pass only dates!');
```

Commit [307c9ce](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/307c9ce718ef66c595751fa7dc9722287bbb9e50)

File: database/users.ts [hunk 18]
```diff
@@ -1,6 +1,6 @@
 import { cache } from 'react';
 import { sql } from '../database/connect';
-import { User } from '../migrations/00006-createTableUsers';
+import { User } from '../migrations/00006-createTableusers';
```
Comment (ProchaLu) [inline database/users.ts]: Test is failing after pull from main. Will update main after Victors lecture and fix the file name

File: app/api/(auth)/register/route.ts [hunk 19]
```diff
@@ -5,7 +5,7 @@ import { NextRequest, NextResponse } from 'next/server';
 import { z } from 'zod';
 import { createSession } from '../../../../database/sessions';
 import { createUser, getUserByUsername } from '../../../../database/users';
-import { User } from '../../../../migrations/00006-createTableUsers';
+import { User } from '../../../../migrations/00006-createTableusers';
```
Comment (ProchaLu) [inline app/api/(auth)/register/route.ts]: Test is failing after pull from main. Will update main after Victors lecture and fix the file name

File: playwright/navigation.spec.ts (hunk 14 above)
Comment (ProchaLu) [reply playwright/navigation.spec.ts]: Will make the changes in a different PR

File: app/animals/[animalId]/page.tsx (hunk 5 above)
Comment (ProchaLu) [reply app/animals/[animalId]/page.tsx]: Refactored the `getDaysUntilNextBirthday()` and now don't have to use `Math.abs()` anymore. 
Commit [65135d8](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/65135d85f23dcc835c67ea6a07416d20f47d07ce)

File: util/date.ts (hunk 8 above)
Comment (ProchaLu) [reply util/date.ts]: Refactored the `getDaysUntilNextBirthday()`
Commit [65135d8](https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/5/commits/65135d85f23dcc835c67ea6a07416d20f47d07ce)

File: util/date.ts (hunk 11 above)
Comment (ProchaLu) [reply util/date.ts]: This code block can be used in the Next.js lecture without a function. This makes it easier for the students to understand the process of creating a function from a code block without making changes. To use this code block in the Next.js lecture, we only need to add the current date to the code. 

```js
  const currentDate = new Date();

  birthDate.setFullYear(currentDate.getFullYear());

  if (birthDate.getTime() < currentDate.getTime()) {
    birthDate.setFullYear(currentDate.getFullYear() + 1);
  }

  const daysUntilNextBirthday =
    birthDate.getDate() === currentDate.getDate() &&
    birthDate.getMonth() === currentDate.getMonth()
      ? 0
      : Math.ceil(
          (birthDate.getTime() - currentDate.getTime()) / (1000 * 60 * 60 * 24),
        );

```
A detailed description of the lecture changes are in the PR description.

File: playwright/navigation.spec.ts (hunk 14 above)
Comment (ProchaLu) [reply playwright/navigation.spec.ts]: Updated the database values in this PR 
- https://github.com/upleveled/next-js-example-fall-2023-atvie/pull/8

File: database/users.ts (hunk 18 above)
Comment (ProchaLu) [reply database/users.ts]: Fixed in main https://github.com/upleveled/next-js-example-fall-2023-atvie/commit/be18676ba598834fddb1e0093801b760f257874a

File: app/api/(auth)/register/route.ts (hunk 19 above)
Comment (ProchaLu) [reply app/api/(auth)/register/route.ts]: Fixed in main https://github.com/upleveled/next-js-example-fall-2023-atvie/commit/be18676ba598834fddb1e0093801b760f257874a

File: app/animals/[animalId]/page.tsx [hunk 20]
```diff
@@ -23,10 +24,17 @@ export default async function AnimalPage(props: Props) {
...
+    new Date(),
+    singleAnimal.birthDate,
+  );
+
   return (
     <div>
       This is a single animal page
       <h1>{singleAnimal.firstName}</h1>
+      <div>Birth date: {formatDate(singleAnimal.birthDate)}</div>
+      <div>Days left until Birthday: {daysUntilNextBirthday}</div>
```
Comment (karlhorky) [inline app/animals/[animalId]/page.tsx]:
```suggestion
      <div>Days left until birthday: {daysUntilNextBirthday}</div>
```

**Review by karlhorky — APPROVED** [review]
